### ISSUE TITLE
Is it possible to search for multiple lines of text?

### ISSUE THREAD

**Issue by domi91c** [eugen0329/vim-esearch#160]
I haven't seen it mentioned anywhere so I'm assuming it's not possible, but is it possible to search for multiple lines of text?

**Comment by eugen0329** [comment]
Technically, yes.

Switch to a regex mode using `<c-r><c-r>` while within the input and insert `line1\nline2` to search across multiple lines.
Ag works out of the box. For rg add `let g:esearch.adapters = {'rg': {'options': '--multiline'}}`.

In case you want to search code duplicates using a fixed string pattern, you can visually select lines for pre-filling, start the search using `<leader>f` and press enter without modifying the selection.

tldr When fixed strings patterns are used, `\n` is recognized as two characters (`\` and `n`). Pressing `<cr>` within the input will submit it, so you must use either regex mode, where `\n` is interpreted as the new line char, or submit a pre-filled text without changes.

**Comment by eugen0329** [comment]
Alternatively, you can write your own preprocessor. It escapes special characters and allows matching lines with any indentation level.

```vim
function! s:preprocess(esearch) abort
  if get(a:esearch, 'search_duplicates', v:false)
    let a:esearch.regex = a:esearch._adapter.bool2regex[v:true] " Force regex mode

    for p in a:esearch.pattern.list
      let p.str = substitute(p.str, '\\n', "\n", 'g')
      let p.str = escape(p.str, '|.+*?()[{^$\') " escape PCRE special chars
      let p.str = substitute(p.str, '[[:blank:]]\+', '[[:blank:]]+', 'g')
    endfor
  endif

  return a:esearch
endfu

au User eseach_config_eager_post call g:esearch.middleware.insert_after('input', function('s:preprocess'))

" Define a separate map where s:preprocess will be used
nnoremap <leader>fd :call esearch#init({'search_duplicates': v:true})<cr>
xnoremap <expr><leader>fd esearch#prefill({'search_duplicates': v:true})
```

**Comment by eugen0329** [comment]
I'm closing it. Please, feel free to ask any questions or to reopen the issue.
